File: crates/libs/lib-core/src/repositories/system.rs
use async_trait::async_trait;
use futures::stream::TryStreamExt;
use mongodb::{
    bson::{doc, oid::ObjectId, Bson, Document},
    Collection,
};

use crate::{errors::CoreError, models::System, Result};

#[derive(Clone)]
pub struct MongoSystemRepository {
    collection: Collection<System>,
}

pub struct CreateSystemDTO {
    pub name: String,
    pub neighbours: Vec<ObjectId>,
}

pub struct UpdateSystemDTO {
    pub name: Option<String>,
    pub neighbours: Option<Vec<ObjectId>>,
}

#[async_trait]
pub trait SystemRepository {
    async fn create(&self, data: CreateSystemDTO) -> Result<ObjectId>;
    async fn find_one(&self, oid: ObjectId) -> Result<Option<System>>;
    async fn find_one_by_name(&self, name: String) -> Result<Option<System>>;
    async fn find_all(&self, filter: Document) -> Result<Vec<System>>;
    async fn delete(&self, oid: ObjectId) -> Result<()>;
    async fn update(&self, oid: ObjectId, data: UpdateSystemDTO) -> Result<()>;
}

impl MongoSystemRepository {
    pub fn new(collection: Collection<System>) -> Self {
        Self { collection }
    }
}

#[async_trait]
impl SystemRepository for MongoSystemRepository {
    async fn create(&self, data: CreateSystemDTO) -> Result<ObjectId> {
        let id = self
            .collection
            .insert_one(System {
                name: data.name,
                neighbours: data.neighbours,
                ..Default::default()
            })
            .await
            .map_err(|_| CoreError::ServerError)?
            .inserted_id
            .as_object_id()
            .ok_or(CoreError::ServerError)?;

        Ok(id)
    }

    async fn find_one(&self, oid: ObjectId) -> Result<Option<System>> {
        let system = self
            .collection
            .find_one(doc! {"_id": oid})
            .await
            .map_err(|_| CoreError::ServerError)?;

        Ok(system)
    }

    async fn find_one_by_name(&self, name: String) -> Result<Option<System>> {
        let system = self
            .collection
            .find_one(doc! {"name": name})
            .await
            .map_err(|_| CoreError::ServerError)?;

        Ok(system)
    }

    async fn find_all(&self, filter: Document) -> Result<Vec<System>> {
        let cursor = self
            .collection
            .find(filter)
            .await
            .map_err(|_| CoreError::ServerError)?;

        cursor
            .try_collect()
            .await
            .map_err(|_| CoreError::ServerError)
    }

    async fn delete(&self, oid: ObjectId) -> Result<()> {
        self.collection
            .find_one_and_delete(doc! {"_id": oid})
            .await
            .map_err(|_| CoreError::ServerError)?;
        Ok(())
    }

    async fn update(&self, oid: ObjectId, data: UpdateSystemDTO) -> Result<()> {
        let mut update = Document::new();

        if let Some(name) = data.name {
            update.insert("name", Bson::String(name));
        }

        if let Some(neighbours) = data.neighbours {
            update.insert(
                "neighbours",
                Bson::Array(neighbours.iter().map(|v| Bson::ObjectId(*v)).collect()),
            );
        }

        self.collection
            .find_one_and_update(doc! {"_id": oid}, doc! {"$set": update})
            .await
            .map_err(|e| {
                eprintln!("{}", e);
                CoreError::ServerError
            })?;

        Ok(())
    }
}

File: api/repositories/system.go
package repositories

import (
	"astragalaxy/models"
	"errors"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type SystemRepository struct {
	db gorm.DB
}

func NewSystemRepository(db gorm.DB) SystemRepository {
	return SystemRepository{db: db}
}

func (r *SystemRepository) Create(s *models.System) (*uuid.UUID, error) {
	m := models.System{
		Name: s.Name,
	}
	if err := r.db.Create(&m).Error; err != nil {
		return nil, err
	}
	return &m.ID, nil
}

func (r *SystemRepository) FindOne(ID uuid.UUID) (*models.System, error) {
	var m models.System

	if err := r.db.Find(&m, ID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}

	return &m, nil
}

func (r *SystemRepository) FindOneByName(name string) (*models.System, error) {
	var m models.System
	if err := r.db.Where(models.System{Name: name}).First(&m).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}

	return &m, nil
}

func (r *SystemRepository) FindAll() []models.System {
	var systems []models.System
	r.db.Find(&systems)

	return systems
}

func (r *SystemRepository) Delete(ID uuid.UUID) error {
	return r.db.Delete(&models.System{}, ID).Error
}

func (r *SystemRepository) Update(s *models.System) error {
	return r.db.Model(&s).Updates(&s).Error
}
